### PR TITLE
Fake fields must not be json encoded when array casted by model, fixes 478

### PR DESCRIPTION
**Pull request by fede91it**

### CODE REVIEW

**Comment by fede91it** [comment]
There is something to fix here?

File: src/PanelTraits/FakeFields.php [hunk 1]
```diff
@@ -66,4 +69,16 @@ public function compactFakeFields($request, $form = 'create', $id = false)
         // since no modifications or additions have been made to $request
         return $request;
     }
+
+    /**
```
Comment (lloy0076) [inline src/PanelTraits/FakeFields.php]: We probably could describe this method/function better :)

File: src/PanelTraits/FakeFields.php [hunk 2]
```diff
@@ -66,4 +69,16 @@ public function compactFakeFields($request, $form = 'create', $id = false)
...
+     * @return array
+     */
+    protected function getModelCasts()
+    {
+        $modelReflection = new \ReflectionClass($this->model);
+        $castsAttribute = $modelReflection->getProperty('casts');
+        $castsAttribute->setAccessible(true);
```
Comment (lloy0076) [inline src/PanelTraits/FakeFields.php]: Should a function conspicuously named **get** be changing the accessibility of an attribute?

Better yet, why?

In fact, if this really does only get the casts it shouldn't be making them accessible - I'd suggest refactoring that out to another method or calling it straight after the `getModelCasts()` above unless there's a compelling reason to keep it here [in which case it probably needs a comment to stop others wondering why `get` functions are running about changing accessible casts].

File: src/PanelTraits/FakeFields.php [hunk 3]
```diff
@@ -66,4 +69,16 @@ public function compactFakeFields($request, $form = 'create', $id = false)
...
+
+        return $castsAttribute->getValue($this->model);
```
Comment (lloy0076) [inline src/PanelTraits/FakeFields.php]: This reads less surprisingly as:

```
    $castedFields = $castsAttribute->getValue($this->model);
    return $castedFields;
```

I know that it's not necessary to split those two out but I'm sitting here wondering, "Hmmm, why does something that **looks like** a plural suddenly just get a single value?" At least breaking it into two makes it more obvious and PHP should be sane enough to optimize the double step out.

**Comment by fede91it** [comment]
I provided a working solution to that problem, without change anything in Backpack.

**Comment by fede91it** [comment]
Any news here? It is from may that the PR is ready 😢

File: src/PanelTraits/FakeFields.php (hunk 1 above)
Comment (lloy0076) [reply src/PanelTraits/FakeFields.php]: Does this have a description?

File: src/PanelTraits/FakeFields.php (hunk 2 above)
Comment (lloy0076) [reply src/PanelTraits/FakeFields.php]: Why is that `setAccessible(true)` there - I'd be inclined to comment it.

**Comment by tabacitu** [comment]
Hi @fede91it thanks a lot for the PR. The code looks very good, as always.

I'm just worried about one thing - whether this is a breaking change or not. Since some apps have been running with double-encoded JSONs for a while, if they run ```composer update``` their apps will stop working - right? If so, I believe we should either:
A) merge this into the ```upgrade``` branch and instruct people to ```json_decode``` their databases (which will be a bitch - we should probably offer a solution to that too);
B) merg this into the ```dev``` branch and automagically ```json_decode``` their databases when they edit an item;

@tumf87 you also mentioned you're interested in this - what do you think about the aboe? Also does @fede91it 's solution work for you? 

Sorry for taking so long to take care of this @fede91it - it's difficult issue so I kept postponing it, but not acknowledging it won't make it go away, I know...

**Comment by lloy0076** [comment]
I'm not sure that the original behaviour is correct...

**Comment by tumf87** [comment]
Hi guys,

So I've done some testing for this issue, _without the changes in this PR_, and I cannot get to reproduce the double encoding part.

What I've done:
- Created a Model with the `HasTranslations` trait
- Added some columns to the `$fillable` attribute, including a text column type for fakes (different name than 'extras')
- Added some of the columns to the `$translatable` attribute, including the fake column
- Added the fake column to the `$casts` attribute and specified array as the cast type
- Added some fake CRUD fields that store in the previously mentioned fake column

Saving and updating the model doesn't seem to double encode the fake field in the database.

Am I missing something, looking in the wrong place?

**Comment by tabacitu** [comment]
@fede91it - I also tested with the current version of Backpack and can't seem to get the model double-encoded. Do you still have this problem after ```composer update```? If so, how can we replicate it?

Thanks, cheers!

**Comment by fede91it** [comment]
Hi @tabacitu, it works fine now!

**Comment by OliverZiegler** [comment]
Isn't this one addressing the same issue fixed in #1116 ?

**Comment by kiddtang** [comment]
@tabacitu  tested with 3.3.3 without the @OliverZiegler 's PR #1116 
(Note: doing `php artisan cache:clear` before each test), 
i observe **no double encoded**. 
However, without cache clearing, somehow it can be happened again. i am not sure why.
By the way, cast to array will cause error, so the document is confusing...
https://laravel-backpack.readme.io/v3.3/docs/advanced-features#section-extras-fake-fields-stored-as-json-in-the-database
```
protected $casts = [
        'fake1' => 'array',
        'fake2' => 'array',
]
```

Although it is not double coded, the fields still unable to retrieved the stored data...
Basically, I just declared `$fakeColumns` and set the `json_decode` in `CrudTrait.php` to give me associative array
```
protected $fakeColumns = ['fake1', 'fake2'];
```
```
$column_contents = json_decode($this->{$column}, true);   // obtain associative arrays
```
Looking forward this issue to be fixed.

**Comment by tabacitu** [comment]
Hi guys. Looks like the issue has been fixed since the PR, since Federico confirmed ([see my tests here too](https://github.com/Laravel-Backpack/CRUD/pull/1116#issuecomment-352710902)), so I'm going to close this.

CRUD 3.3.4 seems not to be plagued by this issue anymore.

Thanks, cheers!
